stl/inc/limits

```diff
@@ -134,7 +138,10 @@ struct _Num_int_base : _Num_base { // base for integer types
 };
 
 struct _Num_float_base : _Num_base { // base for floating-point types
-    static constexpr float_denorm_style has_denorm = denorm_present;
+    _STL_DISABLE_DEPRECATED_WARNING
+    _CXX23_DEPRECATE_DENORM static constexpr float_denorm_style has_denorm = denorm_present;
+    _STL_RESTORE_DEPRECATED_WARNING

```

Comment:
When does this deprecation warning actually get fired? Given `_STL_DISABLE_DEPRECATED_WARNING` unconditionally disables the warning?

Reply:
Good question. The deprecation warning is emitted for any mention of:

* The `enum` type
* The enumerator names (which are available in `namespace std` because this is a classic `enum`)
* The `static constexpr` data members, including this one

The disable/restore machinery prevents the deprecation warnings from being *emitted* within the STL's own headers (because we need to mention the type and the enumerator here). However, it doesn't prevent the "this is deprecated" attribute from being recorded for this `static constexpr` data member, so any mention in *user code* of `has_denorm` will warn. (We follow this same practice elsewhere in the STL where we need to mention our own deprecated stuff.)

Fully worked example:

```
D:\GitHub\STL\out\x64>type meow.cpp
```
```cpp
#include <limits>
using namespace std;

int main() {
    // Mention the type:
    float_denorm_style* ptr = nullptr; // warning C4996: 'std::float_denorm_style': warning STL4042
    (void) ptr;

    // Mention the enumerators:
    auto e1 = denorm_indeterminate; // warning C4996: 'std::float_denorm_style': warning STL4042
    auto e2 = denorm_absent; // warning C4996: 'std::float_denorm_style': warning STL4042
    auto e3 = denorm_present; // warning C4996: 'std::float_denorm_style': warning STL4042
    (void) e1;
    (void) e2;
    (void) e3;

    // Mention the static data members:
    auto e4 = numeric_limits<int>::has_denorm; // warning C4996: 'std::_Num_base::has_denorm': warning STL4042
    bool b5 = numeric_limits<int>::has_denorm_loss; // warning C4996: 'std::_Num_base::has_denorm_loss': warning STL4042
    auto e6 = numeric_limits<float>::has_denorm; // warning C4996: 'std::_Num_float_base::has_denorm': warning STL4042
    (void) e4;
    (void) b5;
    (void) e6;
}
```
```
D:\GitHub\STL\out\x64>cl /EHsc /nologo /W4 /std:c++20 meow.cpp
meow.cpp

D:\GitHub\STL\out\x64>cl /EHsc /nologo /W4 /std:c++latest meow.cpp
meow.cpp
meow.cpp(6): warning C4996: 'std::float_denorm_style': warning STL4042: std::float_denorm_style, std::numeric_limits::has_denorm, and std::numeric_limits::has_denorm_loss are deprecated in C++23. You can define _SILENCE_CXX23_DENORM_DEPRECATION_WARNING or _SILENCE_ALL_CXX23_DEPRECATION_WARNINGS to suppress this warning.
meow.cpp(10): warning C4996: 'std::float_denorm_style': warning STL4042: std::float_denorm_style, std::numeric_limits::has_denorm, and std::numeric_limits::has_denorm_loss are deprecated in C++23. You can define _SILENCE_CXX23_DENORM_DEPRECATION_WARNING or _SILENCE_ALL_CXX23_DEPRECATION_WARNINGS to suppress this warning.
D:\GitHub\STL\out\x64\out\inc\limits(40): note: see declaration of 'std::float_denorm_style'
meow.cpp(11): warning C4996: 'std::float_denorm_style': warning STL4042: std::float_denorm_style, std::numeric_limits::has_denorm, and std::numeric_limits::has_denorm_loss are deprecated in C++23. You can define _SILENCE_CXX23_DENORM_DEPRECATION_WARNING or _SILENCE_ALL_CXX23_DEPRECATION_WARNINGS to suppress this warning.
D:\GitHub\STL\out\x64\out\inc\limits(40): note: see declaration of 'std::float_denorm_style'
meow.cpp(12): warning C4996: 'std::float_denorm_style': warning STL4042: std::float_denorm_style, std::numeric_limits::has_denorm, and std::numeric_limits::has_denorm_loss are deprecated in C++23. You can define _SILENCE_CXX23_DENORM_DEPRECATION_WARNING or _SILENCE_ALL_CXX23_DEPRECATION_WARNINGS to suppress this warning.
D:\GitHub\STL\out\x64\out\inc\limits(40): note: see declaration of 'std::float_denorm_style'
meow.cpp(18): warning C4996: 'std::_Num_base::has_denorm': warning STL4042: std::float_denorm_style, std::numeric_limits::has_denorm, and std::numeric_limits::has_denorm_loss are deprecated in C++23. You can define _SILENCE_CXX23_DENORM_DEPRECATION_WARNING or _SILENCE_ALL_CXX23_DEPRECATION_WARNINGS to suppress this warning.
meow.cpp(19): warning C4996: 'std::_Num_base::has_denorm_loss': warning STL4042: std::float_denorm_style, std::numeric_limits::has_denorm, and std::numeric_limits::has_denorm_loss are deprecated in C++23. You can define _SILENCE_CXX23_DENORM_DEPRECATION_WARNING or _SILENCE_ALL_CXX23_DEPRECATION_WARNINGS to suppress this warning.
meow.cpp(20): warning C4996: 'std::_Num_float_base::has_denorm': warning STL4042: std::float_denorm_style, std::numeric_limits::has_denorm, and std::numeric_limits::has_denorm_loss are deprecated in C++23. You can define _SILENCE_CXX23_DENORM_DEPRECATION_WARNING or _SILENCE_ALL_CXX23_DEPRECATION_WARNINGS to suppress this warning.
```

Reply:
OK I figured it was something like this.